### PR TITLE
update apt sources, add apt keyring gpg

### PR DESCRIPTION
**Pull request by rdxmb**
fix https://github.com/geerlingguy/ansible-role-kubernetes/issues/149
fix https://github.com/geerlingguy/ansible-role-kubernetes/issues/105

see https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl

### CODE REVIEW

**Comment by rdxmb** [comment]
I've searched for a solution without using the `shell` module, but all of them were even more ugly ...

Any other ideas for that?

File: defaults/main.yml [hunk 1]
```diff
@@ -51,10 +51,12 @@ kubernetes_apiserver_advertise_address: ''
...
 
-kubernetes_apt_release_channel: main
-# Note that xenial repo is used for all Debian derivatives at this time.
-kubernetes_apt_repository: "deb http://apt.kubernetes.io/ kubernetes-xenial {{ kubernetes_apt_release_channel }}"
+kubernetes_apt_release_channel: "stable"
+kubernetes_apt_keyring_file: "/etc/apt/keyrings/kubernetes-apt-keyring.gpg" 
+kubernetes_apt_repository_pkgs_k8s_io: "deb [signed-by={{ kubernetes_apt_keyring_file }}] https://pkgs.k8s.io/core:/{{ kubernetes_apt_release_channel }}:/v{{ kubernetes_version }}/deb/ /"
 kubernetes_apt_ignore_key_error: false
+# this one is deprecated and will be deleted with this role
+kubernetes_apt_repository: "deb http://apt.kubernetes.io/ kubernetes-xenial {{ kubernetes_apt_release_channel }}"
```
Comment (rdxmb) [inline defaults/main.yml:59]: this variable is not used anymore and could be removed here

File: tasks/setup-Debian.yml [hunk 2]
```diff
@@ -6,16 +6,26 @@
...
+
+- name: Get Kubernetes apt key.
+  shell: "curl -fsSL https://pkgs.k8s.io/core:/{{ kubernetes_apt_release_channel }}:/v{{ kubernetes_version }}/deb/Release.key | gpg --dearmor -o {{ kubernetes_apt_keyring_file }}"
+  args:
+    creates: "{{ kubernetes_apt_keyring_file }}"
+
+- name: Be sure deprecated Kubernetes repository is absent.
+  file: 
+    path: "/etc/apt/sources.list.d/apt_kubernetes_io.list"
+    state: absent
```
Comment (rdxmb) [inline tasks/setup-Debian.yml:23]: if there is a safe way to do that with `apt_repository`, please let me know

**Comment by geerlingguy** [comment]
This is how I currently solve this: https://www.jeffgeerling.com/blog/2022/aptkey-deprecated-debianubuntu-how-fix-ansible

**Comment by rdxmb** [comment]
@geerlingguy yes, I have already seen that. However, `/etc/apt/trusted.gpg.d/` does not seem to be the recommended way. Most docs recommend to add the signing key explicitly to the source.

The trick seems to be the `.asc` here, so I will change that.

What about the other stuff with deleting the old source?

**Comment by rdxmb** [comment]
ready for review.

**Comment by geerlingguy** [comment]
@rdxmb - I realized my edit to that blog post may have not taken effect; there's a newer/simpler way to do it correctly using Ansible's `ansible.builtin.deb822_repository` module. This only works in Ansible 2.15+, so the role meta version requirement would need to be bumped too if we go that route. But it seems like that would be the cleanest method.

**Comment by rdxmb** [comment]
to be honest, I didn't realize the upgrade on your blog post. 

Maybe I will have a look at `deb822_repository`, at the moment this PR works for my environment.

**Comment by ndouglas** [comment]
I opened a similar PR, #153, because I had a miserable time getting `deb822_repository` to work for me. I finally did, but I want to make clear that my PR is largely untested (I'm not actually using this project) and there is some cleanup work that needs to be done as part of the migration if you're working with an existing cluster.

**Comment by rhino5oh** [comment]
@geerlingguy Can we get either this PR in or make some progress on #153? The current repos have been deprecated since August 2023 and this role has been broken in its current state since January 31st...

**Comment by geerlingguy** [comment]
Merging this for now — I would like to move to `deb822_repository` at some point of course :)

**Comment by rhino5oh** [comment]
Thank you! Can you tag the repo with these changes as well?

**Comment by rbaxende** [comment]
I just realized that this PR did not update the yum repository as well, which should have probably be done at the same time. I can try to work up a PR for that....even one that attempts to use deb822_repository, since it looks like #153 doesn't update the yum repo either

**Comment by rhino5oh** [comment]
I just realized that this PR did not update the yum repository as well, which should have probably be done at the same time. I can try to work up a PR for that....even one that attempts to use deb822_repository, since it looks like https://github.com/geerlingguy/ansible-role-kubernetes/pull/153 doesn't update the yum repo either

**Comment by geerlingguy** [comment]
@rhino5oh - oh haha, that makes sense, and I would gladly merge a PR updating yum/dnf too!

**Comment by geerlingguy** [comment]
I've pushed a commit that should hopefully fix yum.

**Pull request merged**